scipy/optimize/zeros.py

```diff
@@ -206,6 +222,86 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     raise RuntimeError(msg)
 
 
+def _array_newton(func, x0, fprime, args, tol, maxiter, fprime2,
+                  failure_idx_flag=False):
+    """
+    A vectorized version of Newton, Halley, and secant methods for arrays. Do
+    not use this method directly. This method is called from :func:`newton`
+    when ``np.isscalar(x0)`` is true. For docstring, see :func:`newton`.
+    """
+    p = np.asarray(x0)
+    failures = np.ones_like(p, dtype=bool)  # at start, nothing converged
+    if fprime is not None:
+        # Newton-Raphson method
+        for iteration in range(maxiter):
+            fder = np.asarray(fprime(p, *args))
+            zero_der = (fder == 0)
+            # stop iterating if all derivatives are zero
+            if zero_der.all():
+                break
+            fval = np.asarray(func(p, *args))
+            # Newton step
+            dp = fval / fder
+            if fprime2 is not None:
+                fder2 = np.asarray(fprime2(p, *args))
+                dp = dp / (1.0 - 0.5 * dp * fder2 / fder)
+            # use last guess where derivative is zero
+            p = np.where(zero_der, p, p - dp)
+            failures = np.abs(dp) >= tol  # items not yet converged
+            # stop iterating if there aren't any failures, not incl zero der
+            if not failures[~zero_der].any():
+                break
+    else:
+        # Secant method
+        dx = np.finfo(float).eps**0.33
+        p1 = p * (1 + dx) + np.where(p >= 0, dx, -dx)
+        q0 = np.asarray(func(p, *args))
+        q1 = np.asarray(func(p1, *args))
+        for iteration in range(maxiter):
+            zero_der = (q1 == q0)
+            # stop iterating if all derivatives are zero
+            if zero_der.all():
+                p = (p1 + p) / 2.0
+                break
+            # Secant Step
+            dp = q1 * (p1 - p) / (q1 - q0)
+            p = np.where(zero_der, (p1 + p) / 2.0, p1 - dp)
+            failures = np.abs(dp) >= tol  # not yet converged
+            # stop iterating if there aren't any failures, not incl zero der
+            if not failures[~zero_der].any():
+                break
+            p1, p = p, p1
+            q0 = q1
+            q1 = np.asarray(func(p1, *args))
+    if zero_der.any():
+        # secant warnings
+        if fprime is None:
+            nonzero_dp = (p1 != p)
+            # non-zero dp, but infinite newton step
+            zero_der_nz_dp = (zero_der & nonzero_dp)
+            if zero_der_nz_dp.any():
+                rms = np.sqrt(
+                    sum((p1[zero_der_nz_dp] - p[zero_der_nz_dp]) ** 2)
+                )
+                warnings.warn("RMS of %s reached" % rms, RuntimeWarning)
```

Comment:
Since this is new code, I'd suggest switching to the `.format()` method on all these warnings.